src/Compiler/Facilities/AsyncMemoize.fs

```diff
@@ -0,0 +1,611 @@
+
+                return
+                    match msg, cached with
+                    | GetOrCompute _, Some(Completed(result, diags)) ->
+                        Interlocked.Increment &hits |> ignore
+                        diags |> replayDiagnostics diagnosticLogger
+                        Existing(Task.FromResult result)
+                    | GetOrCompute(_, ct), Some(Running(tcs, _, _, _, loggers)) ->
+                        Interlocked.Increment &hits |> ignore
+                        incrRequestCount key
+
+                        ct.Register(fun _ ->
+                            let _name = name
+                            Interlocked.Increment &cancel_ct_registration_subsequent |> ignore
+                            post (key, CancelRequest))
+                        |> saveRegistration key
+
+                        loggers.Add diagnosticLogger
+
+                        Existing tcs.Task
+
+                    | GetOrCompute(computation, ct), None
+                    | GetOrCompute(computation, ct), Some(Job.Canceled _)
+                    | GetOrCompute(computation, ct), Some(Job.Failed _) ->
+                        Interlocked.Increment &started |> ignore
+                        incrRequestCount key
+
+                        ct.Register(fun _ ->
+                            let _name = name
+                            Interlocked.Increment &cancel_ct_registration_original |> ignore
+                            post (key, OriginatorCanceled))
+                        |> saveRegistration key
+
+                        let cts = new CancellationTokenSource()
+
+                        cache.Set(
+                            key.Key,
+                            key.Version,
+                            key.Label,
+                            (Running(TaskCompletionSource(), cts, computation, DateTime.Now, ResizeArray()))
+                        )
+
+                        otherVersions
+                        |> Seq.choose (function
+                            | v, Running(_tcs, cts, _, _, _) -> Some(v, cts)
+                            | _ -> None)
+                        |> Seq.iter (fun (_v, cts) ->
+                            use _ = Activity.start $"{name}: Duplicate running job" [| "key", key.Label |]
+                            //System.Diagnostics.Trace.TraceWarning($"{name} Duplicate {key.Label}")
+                            if cancelDuplicateRunningJobs then
+                                //System.Diagnostics.Trace.TraceWarning("Canceling")
+                                cts.Cancel())
+
+                        New cts.Token
+            })
+
+    let internalError key message =
+        let ex = exn (message)
+        failures.Add(key, ex)
+        Interlocked.Increment &errors |> ignore
+    // raise ex -- Suppose there's no need to raise here - where does it even go?
+
+    let processStateUpdate post (key: KeyData<_, _>, action: StateUpdate<_>) =
+        task {
+            do! Task.Delay 0
```

Comment:
`Task.Delay` with 0 ms is a special case and just returns `Task.CompletedTask`.

`Task.Yield ()`?

Reply:
Whoops, yeah, that's what I wanted. Wonder if it's actually necessary.